File: 0303 - Range Sum Query - Immutable/solution.cpp
class NumArray {
private:
    vector<int> dp;
public:
    NumArray(vector<int>& A) {
        int n = A.size();
        dp.resize(n+1);
        for(int i = 1; i <= n; i++) dp[i] = dp[i-1]+A[i-1];
    }
    
    int sumRange(int l, int r) {
        return dp[r+1]-dp[l];
    }
};

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray* obj = new NumArray(nums);
 * int param_1 = obj->sumRange(left,right);
 */

File: 0303 - Range Sum Query - Immutable/solution.py
from itertools import accumulate
class NumArray:

    def __init__(self, A: List[int]):
        self.dp = list(accumulate(A))

    def sumRange(self, l: int, r: int) -> int:
        return self.dp[r]-(self.dp[l-1] if l else 0)


# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# param_1 = obj.sumRange(left,right)